server/protocols/webadmin/webadmin-mailbox-deleted-message-vault/src/main/java/org/apache/james/webadmin/vault/routes/ExportService.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.james.core.MailAddress;
+import org.apache.james.core.User;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.DeletedMessageZipper;
+import org.apache.james.vault.search.Query;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.annotations.VisibleForTesting;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+class ExportService {
+
+    private class ZippedData {
+        private final long contentLength;
+        private final InputStream inputStream;
+
+        private ZippedData(long contentLength, InputStream content) {
+            this.contentLength = contentLength;
+            this.inputStream = content;
+        }
+    }
+
+    private static final String TEMPORARY_FILE_EXTENSION = ".temp";
+
+    private final BlobExportMechanism blobExport;
+    private final BlobStore blobStore;
+    private final DeletedMessageZipper zipper;
+    private final DeletedMessageVault vault;
+
+    @Inject
+    @VisibleForTesting
+    ExportService(BlobExportMechanism blobExport, BlobStore blobStore, DeletedMessageZipper zipper, DeletedMessageVault vault) {
+        this.blobExport = blobExport;
+        this.blobStore = blobStore;
+        this.zipper = zipper;
+        this.vault = vault;
+    }
+
+    Mono<Void> export(User user, Query exportQuery, MailAddress exportToAddress,
+                      Runnable messageToExportCallback) {
+
+        return matchingMessages(user, exportQuery)
+            .doOnNext(any -> messageToExportCallback.run())
+            .collect(Guavate.toImmutableList())
+            .map(Collection::stream)
+            .flatMap(messages -> Mono.fromCallable(() -> zipData(user, messages)))
+            .flatMap(sneakyThrow(zippedData -> blobStore.save(zippedData.inputStream, zippedData.contentLength)))
+            .flatMap(blobId -> exportTo(user, exportToAddress, blobId))
+            .then();
+    }
+
+    private Flux<DeletedMessage> matchingMessages(User user, Query exportQuery) {
+        return Flux.from(vault.search(user, exportQuery))
+            .publishOn(Schedulers.elastic());
+    }
+
+    private ZippedData zipData(User user, Stream<DeletedMessage> messages) throws IOException {
+        File tempFile = temporaryFile();
+        FileOutputStream fileOutputStream = new FileOutputStream(tempFile);
+
+        zipper.zip(message -> loadMessageContent(user, message), messages, fileOutputStream);
+        return new ZippedData(tempFile.length(), new FileInputStream(tempFile));
+    }
+
+    private File temporaryFile() throws IOException {
+        String tempFileName = UUID.randomUUID().toString();
+        File tempFile = Files.createTempFile(tempFileName, TEMPORARY_FILE_EXTENSION).toFile();
```

Comment:
Maybe integrating `james-vault-export-UUID.temp` would give a clue about what this file is...